src/irc.py

```diff
@@ -139,42 +175,82 @@ def _write(self, path: pathlib.Path, source: str) -> None:
+            f"[[ -f {IRC_BRIDGE_CONFIG_PATH}/irc_passkey.pem ]] || "
+            + "openssl genpkey -out {IRC_BRIDGE_CONFIG_PATH}/irc_passkey.pem "
+            + "-outform PEM -algorithm RSA -pkeyopt rsa_keygen_bits:2048",
+        ]
+        logger.info("Creating PEM file for IRC bridge.")
+        exec_process = subprocess.run(
+           pem_create_command, shell=True, check=True, capture_output=True)
+        )
+        logger.info("PEM create output: %s.", exec_process.stdout)
 
-    def _generate_conf_local(self) -> str:
-        """Generate the content of the irc configuration file.
+    def _generate_app_registration_local(self) -> str:
+        """Generate the content of the app registration file.
 
         Returns:
             A string
         """
-        pass
+        app_reg_create_command = [
+                "/bin/bash",
+                "-c",
+                f"[[ -f {IRC_BRIDGE_REGISTRATION_PATH} ]] || "
+                f"/bin/node /app/app.js -r -f {IRC_BRIDGE_REGISTRATION_PATH} "
+                f"-u http://localhost:{IRC_BRIDGE_HEALTH_PORT} "
+                f"-c {IRC_BRIDGE_CONFIG_PATH} -l {IRC_BRIDGE_BOT_NAME}",
+            ],
+        )
 
-    def _generate_app_registration_local(self) -> str:
-        """Generate the content of the app registration file.
+    def _eval_conf_local(self) -> str:
+        """Generate the content of the irc configuration file.
 
         Returns:
-            A string
+            A string to write to the configuration file
         """
-        pass
-
-    def handle_new_db_relation_data(self) -> str:
+        db_string = self._handle_new_db_relation_data()
+        matrix_sting = self._handle_new_matrix_relation_data()
+        # we ignore the matrix string for now until we have a plugins interface
+        with open(f"{IRC_BRIDGE_CONFIG_PATH}/config.yaml", "w") as f:
+            data = yaml.load(f)
+            db_conn = data["database"]["connectionString"]
+            if db_conn == "" or db_conn != db_string:
+                db_conn = db_string
+            # matrix auth string in here TODO
+
+    def _handle_new_db_relation_data(self) -> str:
         """Handle new DB relation data.
 
         Returns:
-            A string
+            A string that conforms to
+            https://github.com/matrix-org/matrix-appservice-irc/blob/develop/config.sample.yaml#L698
+            connectionString: "postgres://username:password@host:port/databasename"
         """
-        pass
+        if self._charm.database.is_relation_ready:
+            db_data = self._charm.database.get_relation_data()
+            db_connect_string = (
+                "postgres://"
+                f"{db_data['POSTGRES_USER']}:"
+                f"{db_data['POSTGRES_PASSWORD']}@"
+                f"{db_data['POSTGRES_HOST']}:"
+                f"{db_data['POSTGRES_PORT']}/"
+                f"{db_data['POSTGRES_DB']}"
+            )
+            return db_connect_string
+        return ""
```

Comment:
I'm not in favor of managing this kind of case by returning empty string as this should lead to a BlockedState with a specific message.
Better to manage with pydantic validation errors